pkg/identity/generate.go

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package identity
+
+import (
+	"context"
+	"crypto/ecdsa"
+
+	"storj.io/storj/pkg/peertls"
+	"storj.io/storj/pkg/storj"
+)
+
+// GenerateKey generates a private key with a node id with difficulty at least
+// minDifficulty. No parallelism is used.
+func GenerateKey(ctx context.Context, minDifficulty uint16) (
+	k *ecdsa.PrivateKey, id storj.NodeID, err error) {
+	var d uint16
+	for {
+		err = ctx.Err()
+		if err != nil {
+			break
+		}
+		k, err = peertls.NewKey()
+		if err != nil {
+			break
+		}
+		id, err = NodeIDFromECDSAKey(&k.PublicKey)
+		if err != nil {
+			break
+		}
+		d, err = id.Difficulty()
+		if err != nil {
+			break
+		}
+		if d >= minDifficulty {
+			return k, id, nil
+		}
+	}
+	return k, id, storj.ErrNodeID.Wrap(err)
+}
+
+// GenerateKeys continues to generate keys until found returns done == false,
+// or the ctx is canceled.
+func GenerateKeys(ctx context.Context, minDifficulty uint16, concurrency int,
+	found func(*ecdsa.PrivateKey, storj.NodeID) (done bool, err error)) error {
```

Comment:
Would an options struct and/or a type alias for `found` make sense here? It's a pretty big function signature.

Reply:
i like the type alias idea. i'll do that